File: backend/src/main/java/com/smartcampus/booking/service/BookingService.java
package com.smartcampus.booking.service;

import com.smartcampus.booking.dto.BookingRequestDTO;
import com.smartcampus.booking.dto.BookingResponseDTO;
import com.smartcampus.booking.dto.RejectRequestDTO;

import com.smartcampus.booking.dto.BookingSuggestionResponseDTO;
import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;

/**
 * Service interface for managing campus resource bookings.
 */
public interface BookingService {
    
    /**
     * Creates a new booking request. 
     * Throws BookingConflictException with suggestions if a conflict is detected.
     */
    BookingResponseDTO createBooking(BookingRequestDTO requestDTO, String currentUserEmail);

    /**
     * Retrieves all bookings created by the current user.
     */
    List<BookingResponseDTO> getMyBookings(String currentUserEmail);

    /**
     * Retrieves all bookings with optional filtering for administrators.
     */
    List<BookingResponseDTO> getAllBookings(String status, String resource, String date);

    /**
     * Retrieves a specific booking by its unique identifier.
     */
    BookingResponseDTO getBookingById(String id, String currentUserEmail, boolean isAdmin);

    /**
     * Approves a pending booking request.
     */
    BookingResponseDTO approveBooking(String id);

    /**
     * Rejects a pending booking request with a reason.
     */
    BookingResponseDTO rejectBooking(String id, RejectRequestDTO requestDTO);

    /**
     * Cancels an existing booking (must be the creator).
     */
    BookingResponseDTO cancelBooking(String id, String currentUserEmail);

    /**
     * Finds up to 3 alternative resources of the same type that are available for the requested slot.
     */
    BookingSuggestionResponseDTO findAlternativeSuggestions(String resourceId, LocalDate date, LocalTime startTime, LocalTime endTime, Integer expectedAttendees);
}

File: frontend/src/services/bookingService.js
import axiosInstance from '../api/axiosInstance';

export const bookingService = {
  // Create a new booking
  createBooking: async (data) => {
    const response = await axiosInstance.post('/bookings', data);
    return response.data;
  },

  // Get current user's bookings
  getMyBookings: async () => {
    const response = await axiosInstance.get('/bookings/my');
    return response.data;
  },

  // Get ALL bookings (ADMIN only)
  getAllBookings: async () => {
    const response = await axiosInstance.get('/bookings');
    return response.data;
  },

  // Get single booking by ID
  getBookingById: async (id) => {
    const response = await axiosInstance.get(`/bookings/${id}`);
    return response.data;
  },

  // Approve a booking (ADMIN)
  approveBooking: async (id) => {
    const response = await axiosInstance.patch(`/bookings/${id}/approve`);
    return response.data;
  },

  // Reject a booking with reason (ADMIN)
  rejectBooking: async (id, reason) => {
    const response = await axiosInstance.patch(`/bookings/${id}/reject`, { reason });
    return response.data;
  },

  // Cancel a booking (USER)
  cancelBooking: async (id) => {
    const response = await axiosInstance.patch(`/bookings/${id}/cancel`);
    return response.data;
  },

  // Delete a booking entirely (USER - for PENDING/REJECTED/CANCELLED)
  deleteBooking: async (id) => {
    const response = await axiosInstance.delete(`/bookings/${id}`);
    return response.data;
  },
};

export default bookingService;
